drafts/zip-whitelisted-transparent-programs.rst

```diff
@@ -0,0 +1,282 @@
+In some situations, more complex predicates are required. Zcash does not yet have any form
+of shielded programmability, and thus users rely on the script system inherited from
+Bitcoin to encumber transparent outputs. But even this is insufficient as-is for some
+desired use cases. In particular, the Bolt protocol [#zip-draft-bolt]_ requires a custom
+opcode, because the script system does not have the necessary primitives; this requires a
+change to the consensus rules.
+
+Additionally, some off-chain protocols require that transaction IDs be non-malleable in
+order to be secure. Zcash transaction IDs are currently malleable in several ways:
+
+- The ``scriptSig`` field of transparent inputs is not part of the data that is hashed for
+  signing (as the signature is placed in that field), but is part of the data that is
+  hashed to produce the transaction ID. An adversary can modify this field in a variety of
+  ways that do not functionally alter it, but change its binary representation. This is
+  the core transaction malleability problem that Bitcoin mitigated with their Segregated
+  Witness soft fork [#bip-0141]_ (for transactions that opt-in to SegWit-only inputs).
+
+- The Ed25519 signature ``joinSplitSig`` is malleable. The known sources of malleability
+  are:
+
+  - The y-coordinate of the R component of the signature is not required to be canonical.
+  - There are equivalent points for R because its order may be greater than the subgroup
+    order ``ell``.
+
+  These malleability sources correspond to the verification process in ``libsodium``. Note
+  that the public key ``joinSplitPubKey`` is also not required to be canonical, but this
+  is not a source of malleability because the public key is hashed as given in the
+  transaction when verifying the signature.
+
+- RedJubjub signatures, as currently enforced by the consensus rules, are malleable. This
+  is because the signature verification algorithm [#redjubjub]_ multiplies through by the
+  cofactor before the identity check (which ensures that the single and batch verification
+  algorithms agree on the set of valid signatures). An adversary can modify the signature
+  by adding a nonzero point of small order to R, leaving the signature valid but altering
+  its binary representation.
+
+  - TODO: Upon further investigation, it looks like RedJubjub is not actually malleable
+    for the above reason. We are analysing this and will update the proposal when we are
+    certain. It is also possible that Ed25519 might also be non-malleable for a similar
+    reason, but verifying this would require very careful inspection of libsodium; it is
+    simpler to just exclude them.
+
+Shielded programmability and a full transaction ID malleability fix would both require
+significant changes across the ecosystem. In the interim, we propose in this ZIP the
+simplest possible change to the consensus rules that addresses the above concerns with a
+minimal impact on the existing ecosystem.
```

Comment:
This also allows incremental migration away from Script as part of the long-term migration from transparent addresses. 

For example, it's trivial to make a program that replicates the current p2pkh Script semantics while removing malleability (and shaving a few bytes off every txn). So an intermediate step is to add programs for current t-addr use cases, and then deprecate the creation of new p2sh and p2pkh outputs

Reply:
This could even potentially allow you to still spend from such transparent programs after we remove script functionality altogether.